File: 151. Reverse Words in a String/reverse.cpp
#include <bits/stdc++.h>

using namespace std;


class Solution {
public:

void reverseStr(string &s, int left, int right){
while(left < right){
swap(s[left],s[right]);
left++;
right--;
}
}

string reverseWords(string s){


int n =s.size();

//step 1 reverse first string
reverseStr(s,0,n-1);

int start=0, end=0;
while(start<n){

 while(start< n && s[start]==' ')start++;

 if(start>n)break;

 end=start;

 while(end < n && s[end] !=' ')end++;

 //step 2 : reverse word
 reverseStr(s,start,end-1);

 start=end;

}

//step 3: remove extra spaces
int i=0,j=0;

while(j<n){
while(j<n && s[j]==' ')j++;

while(j<n && s[j] != ' ')s[i++]=s[j++];

while(j<n && s[j]==' ')j++;

if(j<n)s[i++]=' ';
}

return s.substr(0,i);



}
};

int main(){
string s;
cin>>s;

Solution obj;
cout<<obj.reverseWords(s);

}

File: 151. Reverse Words in a String/reverse.kt
class Solution{

fun reverseStr(arr: CharArray, l: Int, r: Int){
    var left=l
    var right=r

    while(left<right){
        val temp=arr[left]
        arr[left]=arr[right]
        arr[right]=arr[left
        left++
        right--
    }
 }

fun reverseWords(s: String): String {
        val arr=s.toCharrArray()
        val n=arr.size

        reverseStr(arr,0,n-1)

        var start =0


        while(start<n){
            while(start< n && arr[start]==' ')start++

            if(start>n)break

            var end=start
            while(end<n && arr[end]!=' ')end++

            reverseStr(arr,start,end-1)

            start=end
        }

        var i=0
        var j=0

        while(j<n){
            while(j<n && arr[j]==' ')j++

            while(j<n && arr[j]!=' ')arr[i++]=arr[j++]

            while(j<n && arr[j] == ' ')j++

            if(j<n)arr[i++]= ' '

        }


        return String(arr,0,i)
    }

}

fun main(){

val s=readln()

val obj=Solution()

println(obj.reverseWords(s))

}
